Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2022 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+// MARK: - WorkflowObserver
+
+/// The `WorkflowObserver` protocol provides an interface to receive updates during the runtime's
+/// execution loop. All requirements are optional, and no-ops are provided by default.
+public protocol WorkflowObserver {
+    /// Indicates the start of a `WorkflowSession`, which tracks the life of the underlying `WorkflowNode` used to provide renderings for a given `Workflow` type and rendering key.
+    /// - Parameter session: The `WorkflowSession` that began.
+    func sessionDidBegin(
+        _ session: WorkflowSession
+    )
+
+    /// Marks the end of a `WorkflowSession`, indicating that the corresponding `WorkflowNode` has been removed from the tree of Workflows.
+    /// - Parameter session: The `WorkflowSession` that ended.
+    func sessionDidEnd(
+        _ session: WorkflowSession
```

Comment:
it looks like a few of these don't have access to `state`, and none of them have access to the underlying workflow instance. is there an underlying limitation preventing that? I somewhat assumed `WorkflowSession` would contain some of those things at first glance but realized I was wrong after further reading

Reply:
the session stuff in particular was largely modeled on the analogous concepts/interfaces in the [kotlin repo](https://github.com/square/workflow-kotlin/blob/a39c2883708194c157ba592d2982594a86d6f341/workflow-runtime/src/commonMain/kotlin/com/squareup/workflow1/WorkflowInterceptor.kt#L64-L67).

i think i need a bit of clarification on this thought:

> none of them have access to the underlying workflow instance

by 'underlying workflow instance', what do you mean exactly? in many of them the type conforming to `Workflow` is passed as a parameter for the observer to use.

> is there an underlying limitation preventing that?

for the session lifecycle in particular, the state and corresponding workflow (conforming type instance) could be passed. though there is a question about how to order the `sessionDidBegin` and `workflowDidMakeInitialState`.